ReSharper.FSharp/src/FSharp.Psi.Features/src/CodeStyle/FantomasOptionsPage.fs

```diff
@@ -0,0 +1,350 @@
+namespace JetBrains.ReSharper.Plugins.FSharp.Services.Formatter
+
+open System
+open System.Collections.Generic
+open System.IO
+open System.Runtime.InteropServices
+open JetBrains.Application.Notifications
+open JetBrains.Application.UI.Components
+open JetBrains.Application.UI.Options
+open JetBrains.Application.UI.Options.OptionsDialog
+open JetBrains.Collections.Viewable
+open JetBrains.ProjectModel
+open JetBrains.ProjectModel.NuGet.DotNetTools
+open JetBrains.ReSharper.Plugins.FSharp.Settings
+open JetBrains.ReSharper.Resources.Resources.Icons
+open JetBrains.IDE.UI.Extensions
+open JetBrains.Rider.Model
+open JetBrains.Rider.Model.UIAutomation
+open JetBrains.Threading
+open JetBrains.IDE.UI.Extensions.Validation
+
+[<AutoOpen>]
+module private FantomasVersions =
+    let [<Literal>] MinimalSupportedVersion = "3.2"
+    let [<Literal>] BundledVersion = "4.5.11"
+
+type FantomasValidationResult =
+    | Ok
+    | FailedToRun
+    | UnsupportedVersion
+    | SelectedButNotFound
+
+type FantomasVersion =
+    | Bundled
+    | SolutionDotnetTool
+    | GlobalDotnetTool
+
+type FantomasRunSettings =
+    { Version: FantomasVersion * string
+      Path: VirtualFileSystemPath
+      mutable Status: FantomasValidationResult }
+
+type FantomasDiagnosticNotification =
+    { Event: FantomasValidationResult
+      Version: FantomasVersion
+      FallbackVersion: FantomasVersion }
+
+[<SolutionInstanceComponent>]
+type FantomasDetector(lifetime, settingsProvider: FSharpFantomasSettingsProvider,
+                      dotnetToolsTracker: NuGetDotnetToolsTracker) =
+    let [<Literal>] fantomasToolPackageId = "fantomas-tool"
+    let [<Literal>] fantomasPackageId = "fantomas"
+    let minimalSupportedVersion = Version(MinimalSupportedVersion)
+    let notificationSignal = Signal<FantomasDiagnosticNotification>()
+    let rwLock = JetFastSemiReenterableRWLock()
+
+    let calculateGlobalFantomasToolPath version =
+        Path.Combine(CliFolderPathCalculator.ToolPackageFolderPath, fantomasToolPackageId, version, fantomasToolPackageId, version, "tools", "netcoreapp3.1", "any", "Fantomas.dll")
```

Comment:
Could you split this into getting the directory and getting the tool inside it? Just to make it a bit easier to follow.

Also, any reason not to use `FileSystemPath`-derived types?

Reply:
Is `netcoreapp3.1` the only target framework Fantomas uses?

@nojaf Do you know if it can change in foreseeable future?

Reply:
Yes, in 4.7 we will most likely target `net6.0`.

Reply:
Thanks! That means we should not hard-code it @DedSec256.

Reply:
This code is a temporary workaround and will be removed when we add getting global tool paths to the platform since this logic is more non-trivial and different from how we get paths for local tools.